File: frontend/src/types/dashboardApi.ts
import type { GlueBlock, Period } from './dashboard';

export interface DashboardKpis {
  period_label?: string | null;
  current_period_key?: string | null;
  glues: number;
  sku: number;
  stock?: number | null;
  orders: number;
  sales: number;
}

export interface DashboardMeta {
  has_data: boolean;
  org_name?: string | null;
  file_name?: string | null;
  source_format?: string | null;
  imported_at?: string | null;
  limits?: string | null;
  total_blocks: number;
  total_categories: number;
}

export interface FilterOptions {
  periods: Period[];
  subjects: string[];
  brands: string[];
}

export interface CategorySummary {
  subject: string;
  glues: number;
  sku: number;
  orders: number;
  sales: number;
  stock?: number | null;
}

export interface PaginatedCategories {
  items: CategorySummary[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
  from_index: number;
  to_index: number;
  kpis: DashboardKpis;
}

export interface PaginatedBlocks {
  subject: string;
  items: GlueBlock[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
  from_index: number;
  to_index: number;
  kpis: DashboardKpis;
}

export interface PhotoResolveResponse {
  nm: string;
  url: string | null;
}

export interface DashboardQueryParams {
  periodKey?: string;
  subject?: string;
  brand?: string;
  search?: string;
  page?: number;
  pageSize?: number;
}

export function buildQueryString(params: DashboardQueryParams): string {
  const q = new URLSearchParams();
  if (params.periodKey) q.set('periodKey', params.periodKey);
  if (params.subject) q.set('subject', params.subject);
  if (params.brand) q.set('brand', params.brand);
  if (params.search) q.set('search', params.search);
  if (params.page && params.page > 1) q.set('page', String(params.page));
  if (params.pageSize) q.set('pageSize', String(params.pageSize));
  const s = q.toString();
  return s ? `?${s}` : '';
}

export function parseQueryParams(searchParams: URLSearchParams): Required<
  Pick<DashboardQueryParams, 'periodKey' | 'subject' | 'brand' | 'search'>
> & {
  page: number;
  pageSize: number;
} {
  return {
    periodKey: searchParams.get('periodKey') || '',
    subject: searchParams.get('subject') || '',
    brand: searchParams.get('brand') || '',
    search: searchParams.get('search') || '',
    page: Math.max(1, Number(searchParams.get('page') || '1')),
    pageSize: Math.max(1, Number(searchParams.get('pageSize') || '15')),
  };
}

File: backend/app/schemas/dashboard_api.py
from pydantic import BaseModel, Field

from app.schemas.dashboard_data import GlueBlock, Period


class DashboardKpis(BaseModel):
    period_label: str | None = None
    current_period_key: str | None = None
    glues: int = 0
    sku: int = 0
    stock: float | None = None
    orders: float = 0
    sales: float = 0


class DashboardMeta(BaseModel):
    has_data: bool
    org_name: str | None = None
    file_name: str | None = None
    source_format: str | None = None
    imported_at: str | None = None
    limits: str | None = None
    total_blocks: int = 0
    total_categories: int = 0


class FilterOptions(BaseModel):
    periods: list[Period] = Field(default_factory=list)
    subjects: list[str] = Field(default_factory=list)
    brands: list[str] = Field(default_factory=list)


class CategorySummary(BaseModel):
    subject: str
    glues: int
    sku: int
    orders: float
    sales: float
    stock: float | None = None


class PaginatedCategories(BaseModel):
    items: list[CategorySummary]
    page: int
    page_size: int
    total: int
    total_pages: int
    from_index: int = 0
    to_index: int = 0
    kpis: DashboardKpis


class PaginatedBlocks(BaseModel):
    subject: str
    items: list[GlueBlock]
    page: int
    page_size: int
    total: int
    total_pages: int
    from_index: int = 0
    to_index: int = 0
    kpis: DashboardKpis


class PhotoResolveResponse(BaseModel):
    nm: str
    url: str | None = None


class ChartPoint(BaseModel):
    label: str
    value: float


class OrdersSalesPoint(BaseModel):
    label: str
    orders: float
    sales: float


class DashboardCharts(BaseModel):
    kpis: DashboardKpis
    orders_by_subject: list[ChartPoint] = Field(default_factory=list)
    sales_by_subject: list[ChartPoint] = Field(default_factory=list)
    orders_by_period: list[ChartPoint] = Field(default_factory=list)
    top_brands: list[ChartPoint] = Field(default_factory=list)
    orders_vs_sales: list[OrdersSalesPoint] = Field(default_factory=list)


class ProductTableRow(BaseModel):
    nm: str
    vendor_code: str | None = None
    brand: str | None = None
    subject: str | None = None
    period_key: str | None = None
    period_label: str | None = None
    glue_title: str | None = None
    orders: float | None = None
    sales: float | None = None
    stock: float | None = None
    spp: float | None = None
    ad_ctr: float | None = None


class PaginatedProducts(BaseModel):
    items: list[ProductTableRow]
    page: int
    page_size: int
    total: int
    total_pages: int
    from_index: int = 0
    to_index: int = 0
    kpis: DashboardKpis
